src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -0,0 +1,151 @@
+    private SecretBoxOpener secretBoxOpener;
+
+    private class PrivateEncryptedChannelData {
+        byte[] auth;
+        final String channelData;
+
+        protected PrivateEncryptedChannelData(byte[] auth, String channelData) {
+            this.auth = auth;
+            this.channelData = channelData;
+        }
+
+        public byte[] getAuth() {
+            return auth;
+        }
+
+        public String getChannelData() {
+            return channelData;
+        }
+
+        protected void clearAuthToken(){
+            Arrays.fill(auth, (byte)0);
+        }
+    }
+
+    private PrivateEncryptedChannelData authorizerData;
+
+    public PrivateEncryptedChannelImpl(final InternalConnection connection,
+                                       final String channelName,
+                                       final Authorizer authorizer,
+                                       final Factory factory) {
+        super(channelName, factory);
+        this.connection = connection;
+        this.authorizer = authorizer;
+    }
+
+    /* Base class overrides */
+
+    @Override
+    public void bind(final String eventName, final SubscriptionEventListener listener) {
+
+        if (!(listener instanceof PrivateEncryptedChannelEventListener)) {
+            throw new IllegalArgumentException(
+                    "Only instances of PrivateEncryptedChannelEventListener can be bound " +
+                            "to a private encrypted channel");
+        }
+
+        super.bind(eventName, listener);
+    }
+
+
+    protected void prepareChannel() {
+        final String authResponse = getAuthResponse();
+
+        try {
+            final Map authResponseMap = GSON.fromJson(authResponse, Map.class);
+            final String authKey = (String) authResponseMap.get("auth");
+            final String sharedSecret = (String) authResponseMap.get("shared_secret");
+            final String channelData = (String)authResponseMap.get("channel_data");
+
+            if (authKey == null || sharedSecret == null) {
+                throw new AuthorizationFailureException("Didn't receive all the fields we expected " +
+                        "from the Authorizer, expected an auth token and shared_secret but got: "
+                        + authResponse);
+            } else {
+                authorizerData = new PrivateEncryptedChannelData(authKey.getBytes(), channelData);
+                secretBoxOpener = new SecretBoxOpener(Base64.decode(sharedSecret));
+            }
+
+        } catch (final AuthorizationFailureException e) {
+            throw e; // pass this upwards
+        } catch (final Exception e) {
+            // any other errors need to be captured properly and passed upwards
+            throw new AuthorizationFailureException("Unable to parse response from Authorizer: "
+                    + authResponse, e);
+        }
+    }
+
+    @Override
+    public String toSubscribeMessage() {
+
+        prepareChannel();
```

Comment:
```suggestion
        String auth = authenticate();
```
Relates to this [comment](https://github.com/pusher/pusher-websocket-java/pull/234/files#r400016624). Thanks to it we can simplify the code and not needed to hold on to secret `auth` but only use it within the scope we need it.

Reply:
Also indeed it would be more secure to parse it into an array which we can immediately clear (e.g. fill with zeros) when we no longer needed it (after L109). If there were issues with parsing in into `byte[]` maybe `char[]` would work? I would expect Gson to support it out of the box, however even if not it could be achieved with a simple type adapter.

Reply:
It doesn't look like it, trying to do either `final byte[] sharedSecret = (char[])authResponseMap.get("shared_secret");` or `final byte[] sharedSecret = (char[])authResponseMap.get("shared_secret");` gets me `Authentication failure due to [Unable to parse response from Authorizer:`.  I think this is a bigger issue with our GSON parsing in that we're just using generic objects and trying to convert later.

Reply:
that's just casting so that surely won't work but how about having `AuthResponse` class with the two arrays? that's really how Gson should be used.

Reply:
if that could be achieved we could then change the Base64.decode method to accept an array instead of a string

Reply:
After catching up with Marek this afternoon, it looks like this is a limitation of Gson and that we can't convert to byte or char arrays at this time. We've decided to keep the current implementation as we don't keep a hold of any of the private data inside the channel implementation anymore and garbage collection should be sufficient enough to clear the data when it's not being used.